File: src/scb_check/reporting/render.py
"""Render analysis flags for human-readable CLI output."""

from __future__ import annotations

from pathlib import Path

from scb_check.models import AstGrepHit
from scb_check.models import CloneBlock
from scb_check.models import Flags
from scb_check.tree_walking.models import RuleFinding
from scb_check.tree_walking.models import SymbolIR

_RenderedFlag = tuple[tuple[str, int, int], str]


def render_flags(
    flags: Flags,
    source_lines_by_file: dict[Path, tuple[str, ...]],
    *,
    context_lines: int = 1,
    min_duplicate_lines: int | None = None,
) -> str:
    """Render `flags` with surrounding source controlled by `context_lines`."""
    rendered = [
        *_clone_entries(flags, source_lines_by_file, min_duplicate_lines),
        *_ast_grep_entries(flags, source_lines_by_file, context_lines),
        *_structural_entries(flags, source_lines_by_file, context_lines),
        *_complexity_entries(flags, source_lines_by_file, context_lines),
    ]

    if not rendered:
        return ""

    rendered.sort(key=lambda item: item[0])
    return "\n\n".join(text for _, text in rendered)


def _clone_entries(
    flags: Flags,
    source_lines_by_file: dict[Path, tuple[str, ...]],
    min_duplicate_lines: int | None,
) -> list[_RenderedFlag]:
    entries: list[_RenderedFlag] = []
    clone_sloc_lines_by_file = {
        entry.file: entry.lines for entry in flags.lines.clone_sloc_lines_by_file
    }
    for group in _group_clones(flags.findings.clones):
        anchor = group[0]
        if (
            min_duplicate_lines is not None
            and _clone_line_count(anchor, clone_sloc_lines_by_file)
            < min_duplicate_lines
        ):
            continue
        key = (_display_path(anchor.file), anchor.start_line, 0)
        text = _render_clone_group(
            group,
            source_lines_by_file,
            clone_sloc_lines_by_file,
        )
        entries.append((key, text))
    return entries


def _ast_grep_entries(
    flags: Flags,
    source_lines_by_file: dict[Path, tuple[str, ...]],
    context_lines: int,
) -> list[_RenderedFlag]:
    entries: list[_RenderedFlag] = []
    ordered_ast_hits = dict.fromkeys(
        sorted(
            flags.findings.ast_grep_hits,
            key=lambda hit: (
                hit.file.as_posix(),
                hit.line,
                hit.col,
                hit.end_line,
                hit.end_col,
                hit.rule_id,
                hit.message,
                hit.matched_text,
            ),
        ),
    )
    for hit in ordered_ast_hits:
        key = (_display_path(hit.file), hit.line, 1)
        text = _render_ast_grep(hit, source_lines_by_file, context_lines)
        entries.append((key, text))
    return entries


def _structural_entries(
    flags: Flags,
    source_lines_by_file: dict[Path, tuple[str, ...]],
    context_lines: int,
) -> list[_RenderedFlag]:
    entries: list[_RenderedFlag] = []
    for finding in flags.findings.structural_findings:
        key = (_display_path(finding.file), finding.start_line, 2)
        text = _render_structural_finding(
            finding,
            source_lines_by_file,
            context_lines,
        )
        entries.append((key, text))
    return entries


def _complexity_entries(
    flags: Flags,
    source_lines_by_file: dict[Path, tuple[str, ...]],
    context_lines: int,
) -> list[_RenderedFlag]:
    entries: list[_RenderedFlag] = []
    complexity_groups = (
        (
            flags.findings.high_cc_functions,
            3,
            "erosion",
            "complexity",
            "cyc_complexity",
            "complexity > 10",
        ),
        (
            flags.findings.high_cog_functions,
            4,
            "cog_erosion",
            "cognitive complexity",
            "cog_complexity",
            "cognitive complexity > 10",
        ),
    )
    for symbols, kind_rank, title_prefix, value_label, attr_name, threshold in (
        complexity_groups
    ):
        for symbol in symbols:
            key = (_display_path(symbol.file), symbol.start_line, kind_rank)
            complexity_value = getattr(symbol, attr_name)
            text = _render_complexity_warning(
                symbol,
                source_lines_by_file,
                context_lines,
                title=(
                    f"{title_prefix}: function `{symbol.name}` "
                    f"exceeds {value_label} threshold"
                ),
                detail=(
                    f"{value_label}: {complexity_value}, "
                    f"sloc: {symbol.sloc} "
                    f"(threshold: {threshold})"
                ),
            )
            entries.append((key, text))
    return entries


def _group_clones(
    clones: tuple[CloneBlock, ...],
) -> tuple[tuple[CloneBlock, ...], ...]:
    groups: dict[str, list[CloneBlock]] = {}
    for clone in clones:  # scbc: ignore[verbose-list-append-loop]
        groups.setdefault(clone.group_hash, []).append(clone)

    return tuple(
        tuple(
            sorted(
                group,
                key=lambda clone: (_display_path(clone.file), clone.start_line),
            ),
        )
        for group in groups.values()
    )


def _render_clone_group(
    instances: tuple[CloneBlock, ...],
    source_lines_by_file: dict[Path, tuple[str, ...]],
    clone_sloc_lines_by_file: dict[Path, frozenset[int]],
) -> str:
    anchor = instances[0]
    line_count = _clone_line_count(anchor, clone_sloc_lines_by_file)
    line_number_width = max(len(str(clone.end_line)) for clone in instances)
    pad = " " * line_number_width

    lines = [
        (
            "duplicate-structure: duplicated block "
            f"({line_count} lines, {anchor.instance_count} instances)"
        ),
    ]

    for index, clone in enumerate(instances):
        lines.extend([f"{pad} ┆"] if index else [])
        lines.extend(
            [
                f"{pad} ┌─ {_display_path(clone.file)}:{clone.start_line}",
                f"{pad} │",
                *_clone_body_lines(
                    clone,
                    source_lines_by_file,
                    clone_sloc_lines_by_file,
                    line_number_width,
                ),
                f"{pad} │",
            ],
        )
    return "\n".join(lines)


def _clone_body_lines(
    clone: CloneBlock,
    source_lines_by_file: dict[Path, tuple[str, ...]],
    clone_sloc_lines_by_file: dict[Path, frozenset[int]],
    line_number_width: int,
) -> list[str]:
    source_lines = source_lines_by_file.get(clone.file, ())
    if source_lines:
        return [
            f"{line_number:>{line_number_width}} │ "
            f"{_line_at(source_lines, line_number)}"
            for line_number in _clone_source_line_numbers(
                clone,
                source_lines,
                clone_sloc_lines_by_file,
            )
        ]
    return [
        f"{clone.start_line + offset:>{line_number_width}} │ {text}"
        for offset, text in enumerate(clone.first_lines)
    ]


def _clone_line_count(
    clone: CloneBlock,
    clone_sloc_lines_by_file: dict[Path, frozenset[int]],
) -> int:
    if clone.file not in clone_sloc_lines_by_file:
        return clone.end_line - clone.start_line + 1
    sloc_lines = clone_sloc_lines_by_file[clone.file]
    return sum(
        1 for line in sloc_lines if clone.start_line <= line <= clone.end_line
    )


def _clone_source_line_numbers(
    clone: CloneBlock,
    source_lines: tuple[str, ...],
    clone_sloc_lines_by_file: dict[Path, frozenset[int]],
) -> tuple[int, ...]:
    start_line = max(1, min(clone.start_line, len(source_lines)))
    end_line = max(start_line, min(clone.end_line, len(source_lines)))
    line_numbers = tuple(range(start_line, end_line + 1))
    if clone.file not in clone_sloc_lines_by_file:
        return line_numbers
    sloc_lines = clone_sloc_lines_by_file[clone.file]
    return tuple(line for line in line_numbers if line in sloc_lines)


def _render_ast_grep(
    hit: AstGrepHit,
    source_lines_by_file: dict[Path, tuple[str, ...]],
    context_lines: int,
) -> str:
    message = hit.message.strip() or "matches slop pattern"
    source_lines = source_lines_by_file.get(hit.file, ())
    rendered_lines = _source_line_range(
        source_lines,
        hit.line,
        max(hit.line, hit.end_line),
        context_lines,
    )
    fallback_lines = tuple(
        (hit.line + offset, text)
        for offset, text in enumerate(hit.matched_text.splitlines() or [hit.matched_text])
    )
    return "\n".join(
        [
            f"{hit.severity}[{hit.rule_id}]: {message}",
            *_render_source_block(
                f"{_display_path(hit.file)}:{hit.line}:{hit.col + 1}",
                source_lines,
                rendered_lines,
                fallback_lines=fallback_lines,
            ),
        ],
    )


def _render_structural_finding(
    finding: RuleFinding,
    source_lines_by_file: dict[Path, tuple[str, ...]],
    context_lines: int,
) -> str:
    source_lines = source_lines_by_file.get(finding.file, ())
    rendered_lines = _source_line_range(
        source_lines,
        finding.start_line,
        finding.end_line,
        context_lines,
    )
    return "\n".join(
        [
            f"{finding.rule_id}[{finding.severity.value}]: {finding.message}",
            *_render_source_block(
                f"{_display_path(finding.file)}:{finding.start_line}:{finding.span.start_col + 1}",
                source_lines,
                rendered_lines,
            ),
        ],
    )


def _render_complexity_warning(
    symbol: SymbolIR,
    source_lines_by_file: dict[Path, tuple[str, ...]],
    context_lines: int,
    *,
    title: str,
    detail: str,
) -> str:
    source_lines = source_lines_by_file.get(symbol.file, ())
    rendered_lines = _source_line_range(
        source_lines,
        symbol.start_line,
        symbol.start_line,
        context_lines,
    )
    return "\n".join(
        [
            title,
            *_render_source_block(
                f"{_display_path(symbol.file)}:{symbol.start_line}",
                source_lines,
                rendered_lines,
                detail=detail,
            ),
        ],
    )


def _render_source_block(
    location: str,
    source_lines: tuple[str, ...],
    rendered_lines: range,
    *,
    fallback_lines: tuple[tuple[int, str], ...] = (),
    detail: str | None = None,
) -> list[str]:
    fallback_end = max((line for line, _ in fallback_lines), default=0)
    line_number_width = len(str(max(rendered_lines.stop - 1, fallback_end)))
    pad = " " * line_number_width
    lines = [f"{pad} ┌─ {location}", f"{pad} │"]
    if source_lines or not fallback_lines:
        lines.extend(
            f"{line_number:>{line_number_width}} │ "
            f"{_line_at(source_lines, line_number)}"
            for line_number in rendered_lines
        )
    else:
        lines.extend(
            f"{line_number:>{line_number_width}} │ {text}"
            for line_number, text in fallback_lines
        )

    lines.append(f"{pad} │")
    if detail is not None:
        lines.append(f"{pad} = {detail}")
    return lines


def _source_line_range(
    source_lines: tuple[str, ...],
    start_line: int,
    end_line: int,
    context_lines: int,
) -> range:
    context = max(context_lines, 0)
    first_line = max(1, start_line - context)
    last_line = max(start_line, end_line) + context
    if source_lines:
        last_line = min(last_line, len(source_lines))
    return range(first_line, last_line + 1)


def _line_at(source_lines: tuple[str, ...], line_number: int) -> str:
    if line_number <= 0 or line_number > len(source_lines):
        return ""
    return source_lines[line_number - 1]


def _display_path(path: Path) -> str:
    try:
        return path.resolve().relative_to(Path.cwd().resolve()).as_posix()
    except ValueError:
        return path.as_posix()

File: crates/scb-check/src/render.rs
use std::collections::BTreeMap;
use std::path::Path;

use crate::model::{AstGrepFinding, CloneBlock, Function, Report, StructuralFinding};
use serde_json::json;

pub fn render_json(report: &Report) -> String {
    json!({
        "verbosity": json_number(report.verbosity()),
        "erosion": json_number(report.erosion()),
        "cog_erosion": json_number(report.cog_erosion()),
        "files_scanned": report.files_scanned,
        "total_loc": report.total_loc,
        "verbosity_flagged_loc": report.verbosity_flagged_loc,
        "clone_loc": report.clone_loc,
        "ast_grep_flagged_loc": report.ast_grep_flagged_loc,
        "structural_rule_loc": report.structural_rule_loc,
        "structural_rule_findings": report.structural_rule_findings,
        "total_functions": report.total_functions,
        "high_cc_functions": report.high_cc_functions,
        "high_cog_functions": report.high_cog_functions,
        "total_mass": json_number(report.total_mass),
        "high_cc_mass": json_number(report.high_cc_mass),
        "total_cog_mass": json_number(report.total_cog_mass),
        "high_cog_mass": json_number(report.high_cog_mass),
        "syntax_tree_count": report.syntax_tree_count(),
        "syntax_node_count": report.syntax_node_count(),
        "syntax_by_language": syntax_by_language(report),
    })
    .to_string()
}

fn json_number(value: f64) -> f64 {
    if !value.is_finite() || value.abs() < f64::EPSILON {
        0.0
    } else {
        value
    }
}

pub fn render_human(
    report: &Report,
    min_duplicate_lines: Option<usize>,
    context_lines: usize,
) -> String {
    let mut entries = Vec::new();
    let source_index = SourceIndex::new(report);
    append_clone_entries(&mut entries, report, min_duplicate_lines);
    append_ast_grep_entries(&mut entries, report, &source_index, context_lines);
    append_structural_entries(&mut entries, report, &source_index, context_lines);
    append_complexity_entries(&mut entries, report, &source_index, context_lines);
    entries.sort_by(|left, right| left.key.cmp(&right.key));
    entries
        .into_iter()
        .map(|entry| entry.text)
        .collect::<Vec<_>>()
        .join("\n\n")
}

fn append_clone_entries(
    entries: &mut Vec<RenderedEntry>,
    report: &Report,
    min_duplicate_lines: Option<usize>,
) {
    for group in clone_groups(&report.clones) {
        let line_count = clone_line_count(group.anchor());
        if min_duplicate_lines.is_some_and(|minimum| line_count < minimum) {
            continue;
        }
        entries.push(RenderedEntry {
            key: render_key(&group.anchor().file, group.anchor().start_line, 0),
            text: render_clone_group(&group, line_count),
        });
    }
}

fn append_ast_grep_entries(
    entries: &mut Vec<RenderedEntry>,
    report: &Report,
    source_index: &SourceIndex<'_>,
    context_lines: usize,
) {
    for finding in &report.ast_grep_findings {
        entries.push(RenderedEntry {
            key: render_key(&finding.file, finding.start_line, 1),
            text: render_ast_grep(finding, source_index, context_lines),
        });
    }
}

fn append_structural_entries(
    entries: &mut Vec<RenderedEntry>,
    report: &Report,
    source_index: &SourceIndex<'_>,
    context_lines: usize,
) {
    for finding in &report.structural_findings {
        entries.push(RenderedEntry {
            key: render_key(&finding.file, finding.start_line, 2),
            text: render_structural(finding, source_index, context_lines),
        });
    }
}

fn append_complexity_entries(
    entries: &mut Vec<RenderedEntry>,
    report: &Report,
    source_index: &SourceIndex<'_>,
    context_lines: usize,
) {
    for function in report
        .functions
        .iter()
        .filter(|function| function.is_high_cc())
    {
        entries.push(RenderedEntry {
            key: render_key(&function.file, function.start_line, 3),
            text: render_complexity(function, source_index, context_lines, false),
        });
    }
    for function in report
        .functions
        .iter()
        .filter(|function| function.is_high_cog())
    {
        entries.push(RenderedEntry {
            key: render_key(&function.file, function.start_line, 4),
            text: render_complexity(function, source_index, context_lines, true),
        });
    }
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
struct RenderKey {
    path: String,
    line: usize,
    kind_rank: u8,
}

#[derive(Debug, Clone, PartialEq, Eq)]
struct RenderedEntry {
    key: RenderKey,
    text: String,
}

fn render_key(file: &Path, line: usize, kind_rank: u8) -> RenderKey {
    RenderKey {
        path: display_path(file),
        line,
        kind_rank,
    }
}

fn syntax_by_language(report: &Report) -> serde_json::Value {
    let entries = report.syntax_by_language.iter().map(|summary| {
        (
            summary.language.as_str().to_string(),
            json!({
                "tree_count": summary.tree_count,
                "node_count": summary.node_count,
            }),
        )
    });
    entries.collect::<serde_json::Map<_, _>>().into()
}

fn clone_groups(clones: &[CloneBlock]) -> Vec<CloneGroup> {
    let mut clones = clones.to_vec();
    clones.sort_by(|left, right| {
        (
            left.group_hash.as_str(),
            left.file.as_os_str(),
            left.start_line,
        )
            .cmp(&(
                right.group_hash.as_str(),
                right.file.as_os_str(),
                right.start_line,
            ))
    });

    let mut groups: Vec<CloneGroup> = Vec::new();
    for clone in clones {
        if groups
            .last()
            .is_some_and(|group| group.group_hash() == clone.group_hash)
        {
            if let Some(group) = groups.last_mut() {
                group.rest.push(clone);
            }
        } else {
            groups.push(CloneGroup::new(clone));
        }
    }
    groups.sort_by(|left, right| {
        let left_anchor = left.anchor();
        let right_anchor = right.anchor();
        (left_anchor.file.as_os_str(), left_anchor.start_line)
            .cmp(&(right_anchor.file.as_os_str(), right_anchor.start_line))
    });
    groups
}

#[derive(Debug, Clone, PartialEq, Eq)]
struct CloneGroup {
    anchor: CloneBlock,
    rest: Vec<CloneBlock>,
}

impl CloneGroup {
    const fn new(anchor: CloneBlock) -> Self {
        Self {
            anchor,
            rest: Vec::new(),
        }
    }

    const fn anchor(&self) -> &CloneBlock {
        &self.anchor
    }

    fn group_hash(&self) -> &str {
        &self.anchor.group_hash
    }

    fn iter(&self) -> impl Iterator<Item = &CloneBlock> {
        std::iter::once(&self.anchor).chain(self.rest.iter())
    }
}

fn render_clone_group(group: &CloneGroup, line_count: usize) -> String {
    let anchor = group.anchor();
    let line_number_width = group
        .iter()
        .map(|clone| clone.end_line.to_string().len())
        .max()
        .unwrap_or(1);
    let pad = " ".repeat(line_number_width);
    let mut lines = vec![format!(
        "duplicate-structure: duplicated block ({} lines, {} instances)",
        line_count, anchor.instance_count
    )];
    for (index, clone) in group.iter().enumerate() {
        if index > 0 {
            lines.push(format!("{pad} ┆"));
        }
        lines.push(format!(
            "{pad} ┌─ {}:{}",
            display_path(&clone.file),
            clone.start_line
        ));
        lines.push(format!("{pad} │"));
        for (offset, text) in clone.first_lines.iter().enumerate() {
            lines.push(format!(
                "{:>line_number_width$} │ {}",
                clone.start_line + offset,
                text
            ));
        }
        lines.push(format!("{pad} │"));
    }
    lines.join("\n")
}

const fn clone_line_count(clone: &CloneBlock) -> usize {
    clone.sloc
}

fn render_ast_grep(
    finding: &AstGrepFinding,
    source_index: &SourceIndex<'_>,
    context_lines: usize,
) -> String {
    let message = if finding.message.trim().is_empty() {
        "matches slop pattern"
    } else {
        finding.message.trim()
    };
    let mut lines = vec![format!(
        "{}[{}]: {}\n  --> {}:{}:{}",
        finding.severity,
        finding.rule_id,
        message,
        display_path(&finding.file),
        finding.start_line,
        finding.start_col + 1
    )];
    lines.extend(source_block(
        source_index,
        &finding.file,
        finding.start_line,
        finding.end_line,
        context_lines,
    ));
    lines.join("\n")
}

fn render_structural(
    finding: &StructuralFinding,
    source_index: &SourceIndex<'_>,
    context_lines: usize,
) -> String {
    let mut lines = vec![format!(
        "{}[{}]: {}\n  --> {}:{}",
        finding.rule_id,
        finding.severity,
        finding.message,
        display_path(&finding.file),
        finding.start_line
    )];
    lines.extend(source_block(
        source_index,
        &finding.file,
        finding.start_line,
        finding.end_line,
        context_lines,
    ));
    lines.join("\n")
}

fn render_complexity(
    function: &Function,
    source_index: &SourceIndex<'_>,
    context_lines: usize,
    cognitive: bool,
) -> String {
    let mut lines = if cognitive {
        vec![format!(
            "cog_erosion: function `{}` exceeds cognitive complexity threshold\n  --> {}:{}",
            function.name,
            display_path(&function.file),
            function.start_line,
        )]
    } else {
        vec![format!(
            "erosion: function `{}` exceeds complexity threshold\n  --> {}:{}",
            function.name,
            display_path(&function.file),
            function.start_line,
        )]
    };
    lines.extend(source_block(
        source_index,
        &function.file,
        function.start_line,
        function.start_line,
        context_lines,
    ));
    if cognitive {
        lines.push(format!(
            "  = cognitive complexity: {}, sloc: {} (threshold: cognitive complexity > 10)",
            function.cognitive, function.sloc
        ));
    } else {
        lines.push(format!(
            "  = complexity: {}, sloc: {} (threshold: complexity > 10)",
            function.cyclomatic, function.sloc
        ));
    }
    lines.join("\n")
}

fn source_block(
    source_index: &SourceIndex<'_>,
    file: &Path,
    start_line: usize,
    end_line: usize,
    context_lines: usize,
) -> Vec<String> {
    let Some(source_lines) = source_index.lines(file) else {
        return Vec::new();
    };
    let first_line = start_line.saturating_sub(context_lines).max(1);
    let last_line = end_line
        .max(start_line)
        .saturating_add(context_lines)
        .min(source_lines.len());
    let width = last_line.to_string().len();
    let mut rendered = Vec::new();
    for line_number in first_line..=last_line {
        let text = source_lines
            .get(line_number.saturating_sub(1))
            .map(|line| line.trim_end())
            .unwrap_or_default();
        rendered.push(format!("{line_number:>width$} | {text}"));
    }
    rendered
}

#[derive(Debug)]
struct SourceIndex<'a> {
    by_file: BTreeMap<&'a Path, &'a [String]>,
}

impl<'a> SourceIndex<'a> {
    fn new(report: &'a Report) -> Self {
        let by_file = report
            .source_lines
            .iter()
            .map(|source| (source.file.as_path(), source.lines.as_slice()))
            .collect();
        Self { by_file }
    }

    fn lines(&self, file: &Path) -> Option<&'a [String]> {
        let lines = self.by_file.get(file).copied()?;
        (!lines.is_empty()).then_some(lines)
    }
}

fn display_path(path: &Path) -> String {
    let Ok(cwd) = std::env::current_dir() else {
        return path.display().to_string();
    };
    path.strip_prefix(cwd)
        .unwrap_or(path)
        .to_string_lossy()
        .to_string()
}

#[cfg(test)]
mod tests {
    use super::render_human;
    use crate::test_support::{analyze_dir, test_dir, write};

    #[test]
    fn human_output_orders_mixed_findings_and_renders_context_lines() {
        let root = test_dir();
        let config = root.join("scb-check.toml");
        write(&config, "context = 1\n");
        write(
            &root.join("sample.py"),
            r"
def _identity(value):
    return value

def noisy(items):
    for index in range(len(items)):
        print(items[index])
",
        );

        let report =
            analyze_dir(&root, false, false, Some(&config)).expect("analysis should succeed");
        let rendered = render_human(&report, None, 1);

        let structural_index = rendered
            .find("trivial-wrapper[warning]")
            .expect("structural finding should render");
        let ast_index = rendered
            .find("warning[for-range-len]")
            .expect("ast-grep finding should render");
        assert!(structural_index < ast_index);
        assert!(rendered.contains("1 | def _identity(value):"));
        assert!(rendered.contains("2 |     return value"));
        assert!(rendered.contains("4 | def noisy(items):"));
        assert!(rendered.contains("5 |     for index in range(len(items)):"));
        assert!(rendered.contains("6 |         print(items[index])"));
    }
}
